resource/profile.rs

```diff
@@ -0,0 +1,767 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use std::{
+    collections::HashMap,
+    fmt,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc, RwLock,
+    },
+    time::{Duration, Instant},
+};
+use std::fmt::{Display, Formatter};
+
+use itertools::Itertools;
+
+#[derive(Debug)]
+pub struct TransactionProfile {
```

Comment:
I would like to extend this with an 'OpenProfile' as well, since we have to do the causality check + wait there